airbyte-integrations/connectors/source-zendesk-support/source_zendesk_support/streams.py

```diff
@@ -0,0 +1,486 @@
+        # default value if there is not any headers
+        return 60
+
+    @staticmethod
+    def str2datetime(s):
+        """convert string to datetime object"""
+        if not s:
+            return None
+        return datetime.strptime(s, DATETIME_FORMAT)
+
+    @staticmethod
+    def datetime2str(dt):
+        """convert datetime object to string"""
+        return datetime.strftime(dt.replace(tzinfo=pytz.UTC), DATETIME_FORMAT)
+
+
+class UserSettingsStream(SourceZendeskSupportStream):
+    """Stream for checking of a request token and permissions"""
+
+    def path(self, *args, **kwargs) -> str:
+        return "account/settings.json"
+
+    def next_page_token(self, *args, **kwargs) -> Optional[Mapping[str, Any]]:
+        # this data without listing
+        return None
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """returns data from API"""
+        yield from [response.json().get("settings") or {}]
+
+    def get_settings(self) -> Tuple[Mapping[str, Any], Union[str, None]]:
+        for resp in self.read_records(SyncMode.full_refresh):
+            return resp
+        raise SourceZendeskException("not found settings")
+
+
+class IncrementalBasicEntityStream(SourceZendeskSupportStream, ABC):
+    """basic stream for endpoints where an entity name can be used in a path value
+    https://<subdomain>.zendesk.com/api/v2/<entity_name>.json
+    """
+
+    # max size of one data chunk. 100 is limitation of ZenDesk
+    state_checkpoint_interval = SourceZendeskSupportStream.page_size
+
+    # default sorted field
+    cursor_field = SourceZendeskSupportStream.updated_at_field
+
+    # for partial cases when JSON root name of responses is not equal a entity_type value
+    response_list_name: str = None
+
+    def __init__(self, start_date: str, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # add the custom value for skiping of not relevant records
+        self._start_date = self.str2datetime(start_date) if isinstance(start_date, str) else start_date
+
+    def path(self, *args, **kwargs) -> str:
+        return f"{self.entity_type}.json"
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        """returns a list of records"""
+        self.logger.info(
+            "request activity %s/%s" % (response.headers.get("X-Rate-Limit-Remaining", 0), response.headers.get("X-Rate-Limit", 0))
+        )
+
+        # filter by start date
+        for record in response.json().get(self.response_list_name or self.entity_type) or []:
+            if record.get(self.created_at_field) and self.str2datetime(record[self.created_at_field]) < self._start_date:
+                continue
+            yield record
+        yield from []
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
```

Comment:
this method implementation is only used in `IncrementalBasicExportStream` (`IncrementalBasicUnsortedStream` is defining its own `get_updated_state`, and no other class directly extends this one apart from those two) in which case let's push it down there. It's better to define a method implementation at the lowest shared parent level where it's needed.

Reply:
Done